pkg/p2p/peer/processor.go

```diff
@@ -0,0 +1,92 @@
+package peer
+
+import (
+	"bytes"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer/dupemap"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
+	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/eventbus"
+)
+
+// ProcessorFunc defines an interface for callbacks which can be registered
+// to the MessageProcessor, in order to process messages from the network.
+type ProcessorFunc func(message.Message) ([]*bytes.Buffer, error)
+
+// MessageProcessor is connected to all of the processing units that are tied to the peer.
+// It sends an incoming message in the right direction, according to its topic.
+type MessageProcessor struct {
+	dupeMap    *dupemap.DupeMap
+	processors map[topics.Topic]ProcessorFunc
```

Comment:
This is another form of `CallbackListener` from EventBus.  So now we have P2P sending calls to the upper layer directly (via MessageRouter callbacks) but all other layers (chain/mempool/ ...) are sending msg to the P2P via EventBus.Publish. No consistency here.

Reply:
Not sure if there needs to be consistency there, keep in mind that receiving a message from a single peer is a bit different to gossiping a message to the entire network. The event bus is a simple way for components to basically just put items on a set of ring buffers. 

IMO, in this scenario, the event bus makes more sense due to the multi-cast nature of the message, whereas the processing of a single message is fine to handle directly through a callback from the message processor. We can use CallbackListeners if the consistency is really desired, but to me, it seems like an unnecessary layer of indirection, since we are always speaking to a single component.

Reply:
In contrast, the direct responses that need to be made will result as a return argument from the callback that was called - which alleviates the need for a more complex design, i.e. the router, which needed to be instantiated for every peer, and the router itself needed to instantiate lots of different structs that were specifically used to respond directly to peers, with different topics being routed needing one of two dissemination techniques.

Reply:
The `EventBus` is for fire and forget and support N:M message passing. In the case of `MessageProcessor`, if I understand the design, the `ProcessorFunc` is expected to return a value and therefore it is no asynchrony and the communication is 1:1. This would be modeled through the `RPCBus` rather then plumbing `CallbackListener`, however I do not see the need to complicate the matter by adding the `RPCBus`